src/react/helper/room-promise.ts

```diff
@@ -0,0 +1,151 @@
+import { Logger, Room } from '@ably/chat';
+
+/**
+ * RoomPromise is a wrapper around a promise that resolves to a Room instance.
+ *
+ * It is designed to better integrate into the React lifecycle, and control whether an unmount
+ * function needs to be called depending on where the promise resolution occurs relative to the
+ * component lifecycle.
+ */
+export interface RoomPromise {
+  /**
+   * Returns a function to be called when the component is unmounted. If the room promise has resolved at the time,
+   * of calling, then the unmount function returned by the onResolve callback will be called.
+   *
+   * Multiple calls are no-op.
+   *
+   * This should be used in conjunction with React's useEffect hook to ensure that resources are cleaned up.
+   *
+   * Example usage:
+   *
+   * ```ts
+   * useEffect(() => {
+   *   const roomPromise: RoomPromise;
+   *   return roomPromise.unmount();
+   * }, []);
+   *
+   * @returns A function that should be called when the component is unmounted.
+   */
+  unmount: () => () => void;
+}
+
+/**
+ * A callback that can be returned by the onResolve callback to clean up any resources.
+ */
+type UnmountCallback = () => void;
+
+/**
+ * A callback that is called when the promise resolves to a Room instance.
+ */
+export type RoomResolutionCallback = (room: Room) => UnmountCallback;
+
+/**
+ * Default implementation of RoomPromise.
+ */
+class DefaultRoomPromise implements RoomPromise {
+  private readonly _roomId?: string;
+  private readonly _logger: Logger;
+  private readonly _onResolve: RoomResolutionCallback;
+  private _onUnmount?: UnmountCallback;
+  private _unmounted = false;
+
+  /**
+   * Creates a new DefaultRoomPromise and starts the resolution of the promise.
+   *
+   * @param room  The promise that resolves to a Room instance.
+   * @param onResolve  The callback that is called when the promise resolves to a Room instance.
+   * @param logger  The logger to use for logging.
+   * @param roomId The ID of the room, used for logging.
+   */
+  constructor(room: Promise<Room>, onResolve: RoomResolutionCallback, logger: Logger, roomId?: string) {
+    this._roomId = roomId;
+    this._onResolve = onResolve;
+    this._logger = logger;
+
+    this.mount(room).catch(() => {
+      this._logger.trace('DefaultRoomPromise(); mount error', { roomId: this._roomId });
+    });
+  }
+
+  /**
+   * Wait for the room promise to resolve, then execute the onResolve callback, if the component
+   * has not been unmounted. If the component has not been unmounted, then use the returned
```

Comment:
I think one of the `if the component
has not been unmounted` needs to be removed as it reads a bit odd

Reply:
Yes!